File: data-science-major/labs/course-10-mongodb/13_indexes.js
// Experiment 13 -- Creating and testing single-field and compound indexes.
//
// *** NOT EXECUTED ***
// This is the mongosh script for the lab exam. mongod cannot be installed in
// the verification environment (the Debian repositories that host mongodb-org
// are blocked by the egress policy), so this file has never been run here.
// The query logic is executed and asserted in the matching .py file, through
// mongomock. See notes/sem-4/course-10-document-database/lab.md.

use collegeDB

db.students.createIndex({ dept: 1 })
db.students.createIndex({ dept: 1, "marks.maths": -1 })
db.students.createIndex({ email: 1 }, { unique: true })
db.students.createIndex({ dept: 1 }, { name: "dept_idx" })
db.students.getIndexes()
db.students.totalIndexSize()

// --- MEASURE. Do not guess. -------------------------------------------------
db.students.find({ dept: "DS" }).explain("executionStats")
// stage:              COLLSCAN (bad)  vs  IXSCAN (good)
// totalDocsExamined / nReturned:  1 is ideal, 1000 means the index is wrong

// --- the PREFIX rule: an index on {a,b,c} serves a, a+b, a+b+c only ---------
db.students.createIndex({ dept: 1, year: 1, cgpa: 1 })
db.students.find({ dept: "DS" })                        // uses it
db.students.find({ dept: "DS", year: 4 })               // uses it
db.students.find({ year: 4 })                           // does NOT -- COLLSCAN
db.students.createIndex({ year: 1 })                    // so this is needed too

// --- ESR: Equality, Sort, Range ---------------------------------------------
// Query: dept = "DS", maths > 70, sorted by age
db.students.createIndex({ dept: 1, age: 1, "marks.maths": 1 })
//                        ^equality ^sort   ^range
// A range predicate leaves everything AFTER it unordered, so a sort field
// placed after a range field cannot use the index.

// --- a COVERED query: every field is in the index, so no document is read ---
db.students.createIndex({ dept: 1, name: 1 })
db.students.find({ dept: "DS" }, { _id: 0, dept: 1, name: 1 })
          .explain("executionStats")     // totalDocsExamined: 0

// --- a unique index treats a MISSING field as null, and two nulls collide ---
db.students.createIndex({ email: 1 }, { unique: true })
db.students.insertOne({ name: "A" })       // ok -- email missing
db.students.insertOne({ name: "B" })       // DUPLICATE KEY ERROR
db.students.createIndex({ email: 1 },
  { unique: true, partialFilterExpression: { email: { $exists: true } } })

db.students.dropIndex("dept_1")

File: data-science-major/labs/course-10-mongodb/13_indexes.py
"""Experiment 13 — Single-field and compound indexes.

mongomock records indexes but does not report IXSCAN, so this script asserts
that indexes are CREATED and ENFORCED correctly, and demonstrates the prefix
rule and ESR as tables. Run explain("executionStats") on a real server to see
the plan.
"""
import mongomock
from pymongo.errors import DuplicateKeyError
from fixtures import fresh_db


def creating_and_listing():
    db = fresh_db()
    db.students.create_index("dept")
    db.students.create_index([("dept", 1), ("marks.maths", -1)])
    db.students.create_index("age", name="age_idx")

    names = {i["name"] for i in db.students.list_indexes()}
    assert "_id_" in names, "_id is indexed AUTOMATICALLY"
    assert "dept_1" in names
    assert "age_idx" in names, "a custom name"
    assert any("marks.maths" in n for n in names), names

    db.students.drop_index("dept_1")
    assert "dept_1" not in {i["name"] for i in db.students.list_indexes()}

    print(f"  created and listed {len(names)} indexes, including the automatic _id")


def unique_is_enforced():
    db = fresh_db()
    db.students.create_index("name", unique=True)

    try:
        db.students.insert_one({"_id": 99, "name": "Asha"})
        raise AssertionError("expected a DuplicateKeyError")
    except DuplicateKeyError:
        pass

    db.students.insert_one({"_id": 99, "name": "Unique Name"})
    assert db.students.count_documents({}) == 6

    print("  a unique index rejected a duplicate name and accepted a new one")


def unique_and_missing_fields():
    """A missing field indexes as null, and two nulls collide."""
    db = mongomock.MongoClient().collegeDB
    db.people.create_index("email", unique=True)

    db.people.insert_one({"_id": 1, "name": "A"})          # no email -> null
    try:
        db.people.insert_one({"_id": 2, "name": "B"})      # also null
        raise AssertionError("expected a DuplicateKeyError from two nulls")
    except DuplicateKeyError:
        pass

    print("  a unique index allowed ONE document with no email, then rejected")
    print("       the next -- a missing field indexes as null, and nulls collide")
    print("       fix: partialFilterExpression: { email: { $exists: true } }")


def the_prefix_rule():
    """An index on {a,b,c} serves left-hand PREFIXES only."""
    index = ["dept", "year", "cgpa"]
    cases = [
        (["dept"],                 True,  "a prefix"),
        (["dept", "year"],         True,  "a prefix"),
        (["dept", "year", "cgpa"], True,  "the whole index"),
        (["year"],                 False, "not a prefix -- COLLSCAN"),
        (["cgpa"],                 False, "not a prefix -- COLLSCAN"),
        (["year", "cgpa"],         False, "not a prefix -- COLLSCAN"),
    ]

    def is_prefix(fields):
        return index[:len(fields)] == fields

    print(f"  index {{{', '.join(index)}}}:")
    for fields, expected, why in cases:
        assert is_prefix(fields) == expected, (fields, expected)
        mark = "uses it " if expected else "does NOT"
        print(f"    query on {str(fields):28s} {mark}  ({why})")

    print("       the phone book, sorted by (surname, forename): finding every")
    print("       Kumari is fast; finding every Asha means reading the whole book")


def the_esr_rule():
    """Equality, Sort, Range."""
    query = {"equality": "dept", "sort": "age", "range": "marks.maths"}
    correct = ["dept", "age", "marks.maths"]
    wrong = ["dept", "marks.maths", "age"]

    assert correct.index("age") < correct.index("marks.maths"), \
        "the SORT field must come before the RANGE field"
    assert wrong.index("marks.maths") < wrong.index("age"), \
        "this ordering puts the range first, and the sort cannot use the index"

    print(f"  ESR: query is dept = 'DS', maths > 70, sorted by age")
    print(f"    correct: {{{', '.join(correct)}}}   E, S, R")
    print(f"    wrong:   {{{', '.join(wrong)}}}   the range leaves everything")
    print(f"             after it unordered, so the sort falls back to memory")


def covered_query_fields():
    """Every field in the filter AND the projection must be in the index."""
    index = {"dept", "name"}
    filt = {"dept"}
    proj_bad = {"dept", "name", "_id"}      # _id is returned BY DEFAULT
    proj_good = {"dept", "name"}            # with _id: 0

    assert not (filt | proj_bad) <= index, "_id is not in the index -- NOT covered"
    assert (filt | proj_good) <= index, "with _id: 0 it IS covered"

    print("  covered query needs filter + projection inside the index")
    print("       find({dept}, {dept:1, name:1})        NOT covered -- _id sneaks in")
    print("       find({dept}, {dept:1, name:1, _id:0}) COVERED, totalDocsExamined 0")


def what_indexes_cost():
    db = fresh_db()
    for field in ["dept", "age", "active", "name"]:
        db.students.create_index(field)
    n = len(list(db.students.list_indexes()))
    assert n == 5, f"4 plus _id, got {n}"

    print(f"  {n} indexes means every insert, update and delete maintains {n}")
    print(f"       B-trees. Index what you QUERY, not everything -- the limit")
    print(f"       is 64 per collection, and reaching it means something is wrong")


def main():
    print("Experiment 13 -- Indexes")
    creating_and_listing()
    unique_is_enforced()
    unique_and_missing_fields()
    the_prefix_rule()
    the_esr_rule()
    covered_query_fields()
    what_indexes_cost()


if __name__ == "__main__":
    main()
